File: assignments/app/utils/client_ip.py
from starlette.requests import Request


def get_client_ip(request: Request) -> str | None:
    forwarded = request.headers.get("x-forwarded-for")
    if forwarded:
        raw = forwarded.split(",")[0].strip()
    else:
        raw = request.client.host if request.client else None
    if raw and raw.startswith("::ffff:"):
        raw = raw[len("::ffff:") :]
    return raw

File: zones/src/main/java/ec/edu/espe/zonas/security/ClientIp.java
package ec.edu.espe.zonas.security;

import jakarta.servlet.http.HttpServletRequest;

import org.springframework.web.context.request.RequestContextHolder;
import org.springframework.web.context.request.ServletRequestAttributes;

public final class ClientIp {

    private ClientIp() {
    }

    public static String get() {
        ServletRequestAttributes attrs =
                (ServletRequestAttributes) RequestContextHolder.currentRequestAttributes();
        HttpServletRequest request = attrs.getRequest();
        String forwarded = request.getHeader("X-Forwarded-For");
        String raw = (forwarded != null && !forwarded.isBlank())
                ? forwarded.split(",")[0].trim()
                : request.getRemoteAddr();
        return (raw != null && raw.startsWith("::ffff:"))
                ? raw.substring("::ffff:".length())
                : raw;
    }
}
